app/models/targeting.rb

```diff
@@ -0,0 +1,41 @@
+class Targeting < ActiveRecord::Base
+
+  STATIC_TYPE = 'static_query'
+  DYNAMIC_TYPE = 'dynamic_bookmark'
+  TYPES = {STATIC_TYPE => N_('Static Query'), DYNAMIC_TYPE => N_('Dynamic Bookmark')}
+
+  belongs_to :user
+  belongs_to :bookmark
+
+  has_many :targeting_hosts, :dependent => :destroy
+  has_many :hosts, :through => :targeting_hosts
+  has_many :template_invocations, :dependent => :destroy
+
+  validates :targeting_type, :presence => true, :inclusion => Targeting::TYPES.keys
+  validates :bookmark, :presence => true, :if => "dynamic?"
+
+  before_create :assign_search_query, :if => :static?
```

Comment:
So with this, let's say I create invocation with dynamic targeting using bookmark "hostname ~ user*" and schedule it for tomorrow, one hour after me someone modifies the bookmark to "hostname ~ admin*", the invocation will run the command against admin machines right? IIRC we said that it's OK and we should display warning when editing bookmark that there is some targeting that will be affected. Would it make sense to copy the search query even for dynamic so the bookmark can be changed but targeting remains untouched? Then `resolve_hosts!` shouldn't distinguish between static and dynamic and the only difference would be, when is this method called, either during invocation creation or invocation execution.